ext/compressors/lz4/lz4_compress.c

```diff
@@ -121,40 +136,139 @@ lz4_decompress(WT_COMPRESSOR *compressor, WT_SESSION *session,
     size_t *result_lenp)
 {
 	WT_EXTENSION_API *wt_api;
-	char *compressed_data;
+	LZ4_PREFIX prefix;
 	int decoded;
-	size_t src_data_len;
+	uint8_t *dst_tmp;
+
+	(void)src_len;					/* Unused parameters */
 
 	wt_api = ((LZ4_COMPRESSOR *)compressor)->wt_api;
 
-	/* Retrieve compressed length from start of the data buffer. */
-	src_data_len = *(size_t *)src;
-	if (src_data_len + sizeof(size_t) > src_len) {
-		(void)wt_api->err_printf(wt_api,
-		    session,
-		    "lz4_decompress: stored size exceeds buffer size");
-		return (WT_ERROR);
+	/*
+	 * Retrieve the true length of the compressed block and source and the
+	 * decompressed bytes to return from the start of the source buffer.
+	 */
+	memcpy(&prefix, src, sizeof(LZ4_PREFIX));
+
+	/*
+	 * Decompress, starting after the prefix bytes. Use safe decompression:
+	 * we rely on decompression to detect corruption.
+	 *
+	 * Two code paths, one with and one without a bounce buffer. Our caller
+	 * doesn't store the length of the compressed source and so when using
+	 * raw compression, won't have provided us a large enough buffer, so we
+	 * have to allocate a scratch buffer.
+	 */
+	if (dst_len < prefix.compressed_src) {
+		if ((dst_tmp = wt_api->scr_alloc(
+		   wt_api, session, (size_t)prefix.compressed_src)) == NULL)
+			return (ENOMEM);
+
+		decoded = LZ4_decompress_safe(
+		    (const char *)src + sizeof(LZ4_PREFIX), (char *)dst_tmp,
+		    (int)prefix.compressed_len, (int)prefix.compressed_src);
+
+		if (decoded >= 0)
+			memcpy(dst, dst_tmp, dst_len);
```

Comment:
Can the decompress call ever need more than ```dst_len``` bytes? I suspect there is a subtlty here - in that we are compressing more than we actually needed (past a slot boundary). That's not obvious to me from this code.